CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose

veronikaslc opened the pull request.

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose

marta- commented:
@veronikaslc this doesn't look like it would depend on `CARDS-2020`, why is that the base branch?

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,189 @@
+                    // belonging to a visit
+                    + "  inner join [cards:Form] as visitInformation on visitInformation.subject = dataForm.subject"
+                    + " where"
+                    // link to the correct Visit Information questionnaire
+                    + "  visitInformation.questionnaire = '%1$s'"
+                    // link to the correct Patient Information questionnaire
+                    + "  and dataForm.questionnaire = '%2$s'",
```
The questionnaire can't be Visit information and Patient information at the same time. Based on the lines below, it looks like these were supposed tp be Patient information forms.

marta- replied:
My mistake, I misread the query.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,189 @@
+                // link to the correct Patient Information questionnaire
+                + "  and dataForm.questionnaire = '%2$s'"
+                + "  and dataForm.[jcr:uuid] = '%3$s'"
+                + "  order by visitInformation.time desc",
```
`time` is not a property of the form node, it is an answer to a question in visit information forms, and needs to be pulled from `[cards:Answer]`.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,189 @@
+                // belonging to a visit
+                + "  inner join [cards:Form] as dataForm on visitInformation.subject = dataForm.subject"
+                + " where"
+                // link to the correct Visit Information questionnaire
+                + "  visitInformation.questionnaire = '%1$s'"
+                // link to the correct Patient Information questionnaire
+                + "  and dataForm.questionnaire = '%2$s'"
```
Same comment as above, there can be only one questionnaire. These look like they are supposed to be visit information forms.

marta- replied:
My mistake, I misread the query.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,189 @@
+                + "  visitInformation.questionnaire = '%1$s'"
+                // link to the correct Patient Information questionnaire
+                + "  and dataForm.questionnaire = '%2$s'"
+                + "  and dataForm.[jcr:uuid] = '%3$s'"
```
This condition will make the query return no results. The value being passed is the UUID of a Patient information identified by the previous query, and here we're imposing the condition that the Visit information form should have the same UUID, which contradicts the essence of a UUID.

Instead, if we wish to preserve the overall logic of the implementation, this query should select Visit information forms belonging to a Visit subject that is a descendant of the Patient subject that the `form` passed as a parameter on line 142 belongs to.

marta- replied:
My mistake, I misread the query.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,189 @@
+                "select distinct dataForm.*"
+                    + "  from [cards:Form] as dataForm"
+                    // belonging to a visit
+                    + "  inner join [cards:Form] as visitInformation on visitInformation.subject = dataForm.subject"
```
Visit information forms and Patient information forms don't have the same subject. Patient information forms have a Patient subject, Visit information forms have a Visit subject, which is a descendant of a Patient subject.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,189 @@
+            // Query:
+            final Iterator<Resource> resources = resolver.findResources(String.format(
+                // select the data forms
+                "select distinct dataForm.*"
```
The query might be easier to read if we use `patientInformation` instead of `dataForm`.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,189 @@
+            "select distinct visitInformation.*"
+                + "  from [cards:Form] as visitInformation"
+                // belonging to a visit
+                + "  inner join [cards:Form] as dataForm on visitInformation.subject = dataForm.subject"
```
Same comments re joining on `subject` and renaming `dataForm` to `patientInformation` as above.

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Fixed sql query;
Better sql query readability

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,197 @@
+            "select distinct visitInformation.*"
+                + "  from [cards:Form] as visitInformation"
+                // belonging to a visit
+                + "  inner join [cards:Form] as patientInformation on"
+                + "   isdescendantnode(visitInformation.subject, patientInformation.subject)"
```
This join can be avoided, we can just put a `where` condition on `patientInformation.relatedSubjects = '$ param'`, where `param` resolves to the patient information form's subject.

veronikaslc replied:
Why not `visitInformation.subject = `  to the patient information form's subject?

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,197 @@
+            .getResource("/Questionnaires/Visit information/time").getValueMap().get("jcr:uuid");
+        // run query to get all associated Visit information forms sorted by time property
+        Iterator<Resource> resources = resolver.findResources(String.format(
+            // select the data forms
```
All these comments need to be updated to reflect the actual query, both in this and in the query above.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,197 @@
+            return false;
+        }
+        // get the last visit and see if it's surveys_submitted = true
+        final Node visit = resources.next().adaptTo(Node.class);
```
I'm not sure checking just the latest visit is enough. As I understand, there will be a clinic where we send the invitation email 6 months after the visit, so it is possible to have a patient with a "6months later invite" visit from 5 months ago, which is still unsubmitted (and not even invited yet), and a "7days later invite" visit from 8 days ago that has been submitted. In this case, we will still need the patient's details in one month to send the late invitation.

marta- replied:
The Integrated Care surveys would be sent out after 3 months. It's indeed a very different timeline than for all others, however:
* we are not sending more than 1 invite every 6 months
* I believe we had decided that we won't even import any visits that are not viable for invites

Given this ☝️ , I believe we're probably safe looking at the latest visit, but I may be wrong. @sdumitriu ?

sdumitriu replied:
Sounds good.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,197 @@
+                + "  and patientInformation.questionnaire = '%2$s'"
+                + "  and patientInformation.[jcr:uuid] = '%3$s'"
+                // link to the time question
+                + "  and time.question = '%3$s'"
```
Both of these can't be using `%3` at the same time...

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,197 @@
+            Query.JCR_SQL2);
+
+        if (!resources.hasNext()) {
+            return false;
```
If there are no visit information forms, I think we should delete the patient info, since there's no visit to notify about.

veronikaslc replied:
@marta- please confim

marta- replied:
I confirm.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,197 @@
+        }
+    }
+
+    private boolean canDeleteInformation(final Node form, final ResourceResolver resolver, final String visitQ,
```
After looking through the rest of the method, it looks like this is looking for a reason to delete the patient info, but it should be looking for a reason NOT to delete it.

veronikaslc replied:
What is the difference? (if we construct the code correctly)

sdumitriu replied:
In general:

Looking for a reason to do something:
```
if (test) return true;
if (other test) return true;
finally: return false;
```

Looking for a reason NOT to do something:
```
if (test) return false;
if (other test) return false;
finally: return true;
```

The code used to be structured like the former, bun now after the last commit it looks like the latter, with the bug that the final return (and thus all of them) is also false.

veronikaslc replied:
On refactoring now, please don't look yet :)

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,197 @@
+        resources = resolver.findResources(String.format(
+            "select * from [cards:Token]"
+                + " where"
+                + " [" + CardsTokenImpl.TOKEN_ATTRIBUTE_EXPIRY + "] < '%1$s'"
```
This won't work, since tokens are also deleted when they expire, so the chance of finding an expired token is tiny.

marta- commented:
@veronikaslc The specs to https://phenotips.atlassian.net/browse/CARDS-2026 have been updated to make use of the implementation of the `Survey events` form.

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupScheduler.java:
```diff
@@ -0,0 +1,81 @@
+ * Schedule the cleanup of Patient information data every midnight.
+ *
+ * @version $Id$
+ * @since 0.9.2
```
Update `@since`

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Fixed version and query args

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Only remove the value from the answer, not the whole answer node

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Refactored code wrt changed issue description:
First fetch all Patient information forms for which all Servey events have non null discharged_date
Then for each Patient information form fetch all associated Servey event forms in desc visit order
Look at the latest Servey event form to see if it has at least one non null of responses_received or reminder2_sent answers

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,220 @@
+            return false;
+        }
+
+        return false;
```
Seems that all the returns are `false` now.

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Refactored to delete pationt information if there is no Survey events form
Fixed always returning false

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Search only for the patient information forms that do have any private information to remove

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,226 @@
+        final String responsesReceived = (String) resolver
+            .getResource("/Questionnaires/Survey events/responses_received").getValueMap().get(UUID_KEY);
+        final String reminderSent = (String) resolver
+            .getResource("/Questionnaires/Survey events/reminder2Sent").getValueMap().get(UUID_KEY);
```
```suggestion
            .getResource("/Questionnaires/Survey events/reminder2_sent").getValueMap().get(UUID_KEY);
```

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose

Fixed typo in reminder2_sent field

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Bugfix: checked in forms cannot be cleared

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,238 @@
+/**
+ * Periodically clears last_name, first_name, email from any Patient information form of patient subjects
+ * for whom Survey events form corresponding to the latest visit (i.e. for which the answer to discharged_date
+ * is the latest) has at least one of responses_received or reminder2_sent answers filled in with a date value.
```
Comments should say why, not how. We're clearing the information because we no longer need it (why). We determine that we no longer need it by looking at survey events (how). This is an implementation detail that can easily change, making the comment obsolete, but the reason why we do it should remain the same.

sdumitriu committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Speed up the query

sdumitriu committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Reduced cyclomatic complexity

sdumitriu committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Add configuration for enabling/disabling the task

sdumitriu commented:
I've added a configuration for disabling this, and I also moved the `draftLifetime` setting in there. Can @veronikaslc add a new admin config screen for it? Also move the `draftLifetime` config from `PatientAccessConfiguration.jsx` to the new screen.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupScheduler.java:
```diff
@@ -0,0 +1,86 @@
+ * Schedule the cleanup of Patient information data every midnight.
+ *
+ * @version $Id$
+ * @since 0.9.6
```
This needs to be updated now.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,251 @@
+ * is the latest) has at least one of responses_received or reminder2_sent answers filled in with a date value.
+ *
+ * @version $Id$
+ * @since 0.9.6
```
Update.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/PatientInformationCleanupTask.java:
```diff
@@ -0,0 +1,251 @@
+                surveyEventsQuestionnaire, dischargedDate, visitSubjectUuid),
+            Query.JCR_SQL2);
+
+        // If there are no visit information forms, delete the patient info since there's no visit to notify about
```
Wrong comment, the query returned survey events forms, not visit information forms.

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Fixed comments

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/DataRetentionConfigurationImpl.java:
```diff
@@ -0,0 +1,127 @@
+public class DataRetentionConfigurationImpl extends AbstractNodeUtils implements DataRetentionConfiguration
+{
+    /** The location of the configuration node for patient auth. */
+    private static final String CONFIG_NODE = "/lib/cards/conf/DataRetention";
```
@sdumitriu typo in the path, should be `/libs/cards/conf/DataRetention`, unless we need a separate `/lib/`

veronikaslc replied:
Moreover, we never put nodes in the `/cards/conf/`, only json static config files . Are you sure the path is right?
Please see the last commit, where I followed the PatientAccess pattern in data modeling

sdumitriu replied:
The Sling convention is that `/libs/` and `/apps/` are for configurations, with the latter one overriding the former. We do put configuration in both of these, for example the logo settings, the clarity mapping, etc, and although we don't have nice UIs for changing them, they are still configurations. We use `/Survey` as a special place for some settings because they also need to be accessible by guests.

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Added a new data retension admin config screen

sdumitriu committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Change the configuration path to /apps/cards/config/DataRetention
Minor comment fixes

sdumitriu commented on modules/patient-portal/src/main/frontend/webpack.config.js:
```diff
@@ -19,6 +19,8 @@ module.exports = {
     [module_name + 'PatientAccessConfiguration']: { 'dependOn': ['cards-login.loginDialogue'], 'import': './src/patient-portal/PatientAccessConfiguration.jsx' },
     [module_name + 'PatientAccessConfigurationIcon']: '@mui/icons-material/MedicalInformation.js',
+    [module_name + 'DataRetentionConfiguration']: { 'dependOn': ['cards-login.loginDialogue'], 'import': './src/patient-portal/DataRetentionConfiguration.jsx' },
+    [module_name + 'DataRetentionConfigurationIcon']: '@mui/icons-material/MedicalInformation.js',
```
Which one would work better:
- AccessTime
- ContentPasteOff
- DeleteSweep
- BrowseGallery
- Timer
- RemoveDone
- An overlay of MedicalInformation and AccessTime
- An overlay of ContentPasteOff and AccessTime
- something else?

veronikaslc replied:
Picked Timelapse

sdumitriu commented on modules/patient-portal/src/main/frontend/src/patient-portal/DataRetentionConfiguration.jsx:
```diff
@@ -0,0 +1,144 @@
+  // Boolean fields can have one label
+  // Text fields can have one label, one optional helper text, one optional error text
+  const LABELS = {
+    deleteUnneededPatientDetails: "Whether or not unneeded PII should be deleted",
```
```suggestion
    deleteUnneededPatientDetails: "Delete patient's personal information when it is no longer needed",
```

sdumitriu commented on modules/patient-portal/src/main/frontend/src/patient-portal/DataRetentionConfiguration.jsx:
```diff
@@ -0,0 +1,144 @@
+  // Text fields can have one label, one optional helper text, one optional error text
+  const LABELS = {
+    deleteUnneededPatientDetails: "Whether or not unneeded PII should be deleted",
+    deleteDraftAnswers: "Whether or not draft answers should be deleted",
```
```suggestion
    deleteDraftAnswers: "Delete unsubmitted answers",
```

sdumitriu replied:
```suggestion
    deleteDraftAnswers: "Delete unsubmitted answers",
```

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purposeUpdate modules/patient-portal/src/main/frontend/src/patient-portal/DataRetentionConfiguration.jsx

Better wording for labels

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
CARDS-2026: Automatically delete names and emails once they are no longer serving any purpose
Fixed whitespacing and React formatting
